Sources/Workspace/Workspace.swift

```diff
@@ -1199,8 +1210,14 @@ extension Workspace {
         let allManifests = try! topologicalSort(inputManifests.map({ KeyedPair(($0, ProductFilter.everything), key: $0.name)})) { node in
             return node.item.0.dependenciesRequired(for: node.item.1).compactMap({ dependency in
                 let url = config.mirroredURL(forURL: dependency.declaration.url)
-                let manifest = loadedManifests[url] ?? loadManifest(forURL: url, diagnostics: diagnostics)
-                loadedManifests[url] = manifest
+                var manifest = loadedManifests[url]
+                if manifest == nil {
+                    let kind: PackageReference.Kind = (dependency.declaration.requirement == .localPackage) ? .local : .remote
+                    delegate?.willLoadPackage(url: url, kind: kind)
+                    manifest = loadManifest(forURL: url, diagnostics: diagnostics)
+                    delegate?.didLoadPackage(url: url, kind: kind, manifest: manifest, diagnostics: diagnostics.diagnostics)
```

Comment:
Not for this PR but it might also be useful to also put similar callbacks in the package graph loader loop as that operation might be more costly when the manifests are cached: https://github.com/apple/swift-package-manager/blob/master/Sources/PackageGraph/PackageGraphLoader.swift#L184

Reply:
Interesting — thanks for the pointer.  I'm definitely looking for where to put the right calls so that there will be a minimal set of easily explainable, orthogonal callbacks with semantically clear meanings that clients can make good use of.  There's currently a lot of detailed information available at the end of the package graph loading, but having it available as the processing is going on will help all clients present better status and progress information, I think.  I will look closer at this and see if that should be instead of these callbacks, but I will definitely at least change the names as you suggest.  Thanks!